### PR TITLE
contrib/labstack/echo.v4: add appsec integration

### PR DESCRIPTION
**Pull request by Julio-Guerra**
Enable AppSec in echo.v4 by including the AppSec HTTP instrumentation in
the echo middleware function.

### CODE REVIEW

File: contrib/labstack/echo.v4/echotrace.go
Comment (Julio-Guerra) [inline contrib/labstack/echo.v4/echotrace.go:28]: Unrelated change but I though it would be ok to do it here.
Please let me know if you prefer a separate PR.

File: internal/appsec/dyngo/instrumentation/httpinstr/http.go
Comment (gbbr) [inline internal/appsec/dyngo/instrumentation/httpinstr/http.go:69]: This seems unrelated to the PR. Didn't we already do this elsewhere?

File: contrib/labstack/echo.v4/echotrace.go
Comment (gbbr) [reply contrib/labstack/echo.v4/echotrace.go:28]: Fine by me, but can we please remove the empty line? nitty nit

File: internal/appsec/dyngo/instrumentation/httpinstr/http.go
Comment (gbbr) [inline internal/appsec/dyngo/instrumentation/httpinstr/http.go:63]: Can you please leave your Github username? Check point 5 in https://github.com/DataDog/dd-trace-go/wiki/Style-guidelines#general-guidelines

File: internal/appsec/dyngo/instrumentation/httpinstr/http.go
Comment (gbbr) [inline internal/appsec/dyngo/instrumentation/httpinstr/http.go:121]: Same here, would be nice to use your Github username

File: internal/appsec/dyngo/instrumentation/httpinstr/http.go
Comment (gbbr) [inline internal/appsec/dyngo/instrumentation/httpinstr/http.go:115]:
```suggestion
		Host:       host,
```
IMHO you should either do this, or drop the `host` variable altogether. I prefer the latter personally but totally up to you!

File: contrib/labstack/echo.v4/echotrace.go
Comment (gbbr) [inline contrib/labstack/echo.v4/echotrace.go:54]: The problem with your small refactor below is that now the code inside `httpinstr.MakeHandlerOperationArgs` runs every time, adding extra computation even when AppSec is off. Can you fix that somehow please?

**Review by gbbr — CHANGES_REQUESTED** [review]
Left a few thoughts.

File: internal/appsec/dyngo/instrumentation/httpinstr/http.go
Comment (gbbr) [inline internal/appsec/dyngo/instrumentation/httpinstr/http.go:120]: Why not take the `r.Headers` field as it is here and use the same type (`http.Header`) ? Same for the cookies. It would be more standardised and you'd avoid creating new maps and loops.

File: contrib/labstack/echo.v4/echotrace.go
Comment (Julio-Guerra) [reply contrib/labstack/echo.v4/echotrace.go]: Fixed

File: contrib/labstack/echo.v4/echotrace.go
Comment (Julio-Guerra) [reply contrib/labstack/echo.v4/echotrace.go:54]: I did something bold in https://github.com/DataDog/dd-trace-go/pull/1042/commits/01ab93d31dd9894ae438f25eb509ea375fffaf90 which is just adding more `if enabled` guards in `httpinstr` into `StartOperation`, `MakeHandlerOperationArgs` and  `MakeHandlerOperationRes`.

I thought something more advanced with interfaces to lazily build the `HandlerOperationArgs` and `HandlerOperationRes` would be too much for now. I'd rather wait for more integrations before "optimizing" this.

File: internal/appsec/dyngo/instrumentation/httpinstr/http.go
Comment (Julio-Guerra) [reply internal/appsec/dyngo/instrumentation/httpinstr/http.go]: Fixed

File: internal/appsec/dyngo/instrumentation/httpinstr/http.go
Comment (Julio-Guerra) [reply internal/appsec/dyngo/instrumentation/httpinstr/http.go:69]: It is related: this was in the removed lines below and moved out due to the refactoring with `MakeHandlerOperationArgs`.

As Kyle suggested in #1007, we should properly implement these via a new tracer configuration value for global service entry tags. I'll try to do it asap in a separate PR. Until then, this echo.v4 integration won't have those tags.

File: internal/appsec/dyngo/instrumentation/httpinstr/http.go
Comment (Julio-Guerra) [reply internal/appsec/dyngo/instrumentation/httpinstr/http.go:123]: Fixed

File: internal/appsec/dyngo/instrumentation/httpinstr/http.go
Comment (Julio-Guerra) [reply internal/appsec/dyngo/instrumentation/httpinstr/http.go]: Fixed

File: internal/appsec/dyngo/instrumentation/httpinstr/http.go
Comment (gbbr) [inline internal/appsec/dyngo/instrumentation/httpinstr/http.go:144]: This function just adds another layer of complexity. Just writing `HandlerOperationRes{Status: status}` everywhere should be equally good and much clearer.

File: contrib/labstack/echo.v4/echotrace.go
Comment (gbbr) [reply contrib/labstack/echo.v4/echotrace.go:54]: What do you think about having something like `appsec.Enabled()` or `appsec.Status() != appsec.StatusDisabled` (the latter being more appropriate in cases where there would be more statuses than on/off). And simply not running the code at all when it's off, instead of trying to make no-op's out of everything. This no-op pattern doesn't seem to work well in your case.

Comment (Julio-Guerra) [reply contrib/labstack/echo.v4/echotrace.go]: Agreed, much better.
Now fixed.

File: internal/appsec/dyngo/instrumentation/httpinstr/http.go
Comment (Julio-Guerra) [reply internal/appsec/dyngo/instrumentation/httpinstr/http.go]: Reverted, now that we no longer need to no-op it.

File: contrib/labstack/echo.v4/echotrace.go
Comment (gbbr) [inline contrib/labstack/echo.v4/echotrace.go:55]: Can we apply this pattern everywhere now? Just for being consistent.

File: contrib/labstack/echo.v4/echotrace.go
Comment (gbbr) [inline contrib/labstack/echo.v4/echotrace.go:56]: I'm starting to feel quite strongly about renaming this package to `httpsec`. Reading code in the tracer now becomes ambiguous when seeing things like `httpinstr.WrapHandler`. It makes one think "HTTP instrumentation", which in this context, is tracing. We should be clear. I think `httpsec` is a great name ("HTTP Security") and we should probably use this `*sec` suffix on all AppSec integrations, just like we use the `*trace` suffix on all tracing integrations (granted its being renamed at import time but too late to change that).

I have suggested this before. Is there any reason you are against it?

Comment (Julio-Guerra) [reply contrib/labstack/echo.v4/echotrace.go:56]: It's still too early to say this is completely appsec-specific for me. I see some functions currently not covered by spans, like the request body reads, that appsec will need to cover it soon with some new `httpinstr` operations and this will be the moment where the tracer could also create a span as an instrumentation gateway operation callback - because this will be our first time a tracer integration doesn't exist for that. In the base architecture idea, this instrumentation gateway was meant to be used by anyone, this is why I still would like to wait bit and see if it ends up only applied to appsec.

File: contrib/labstack/echo.v4/echotrace.go
Comment (Julio-Guerra) [reply contrib/labstack/echo.v4/echotrace.go:55]: I don't see other spots where it's applicable. I replaced `Status()` by `Enabled()` so that it would be applied everywhere `Status()` was used.

Comment (gbbr) [reply contrib/labstack/echo.v4/echotrace.go:55]: `contrib/internal/httputil/trace.go`

File: contrib/labstack/echo.v4/echotrace.go
Comment (gbbr) [reply contrib/labstack/echo.v4/echotrace.go:56]: Right. But it is part of the `appsec` package, so it is fully AppSec related for now. Would you mind doing this change? I think it offers a much better experience. If we ever end up finding that we need to change the name, it is of course perfectly fine to do so. For now, nothing else is using this for any other purposes than AppSec, and it clashes in concept with HTTP instrumentation of the tracer.

File: contrib/labstack/echo.v4/echotrace.go
Comment (Julio-Guerra) [reply contrib/labstack/echo.v4/echotrace.go:55]: 👍
Done in https://github.com/DataDog/dd-trace-go/pull/1042/commits/c5b4fb03d4d7de82663b66260b1ece5898ed2e7e

I'll add a proper support for global service entry span tags in a separate PR to properly have the `_dd.appsec.enabled: 0` case. The most important part is having it when it's `1` for now so that the service appears in the service list of the appsec dashboard.

File: ddtrace/tracer/log.go
Comment (gbbr) [inline ddtrace/tracer/log.go:109]: Can't this just be a boolean?
```suggestion
		AppSec:                appsec.Enabled(),
```
It would simplify things... Do we have any reason for it to be a string?

File: internal/appsec/appsec.go
Comment (gbbr) [inline internal/appsec/appsec.go:38]: Can we avoid using `unsafe`? If you want, use a plain `uint16` or something, and have `0` means off, and anything else mean on.

File: contrib/labstack/echo.v4/echotrace.go
Comment (gbbr) [inline contrib/labstack/echo.v4/echotrace.go:61]: Can you please remove this empty line?

File: contrib/internal/httputil/trace.go
Comment (gbbr) [inline contrib/internal/httputil/trace.go:61]: And here.

**Review by gbbr — COMMENTED** [review]
Thanks a lot Julio. A few more nits.

File: internal/appsec/appsec.go
Comment (Julio-Guerra) [reply internal/appsec/appsec.go]: Fixed by using a RWMutex instead in https://github.com/DataDog/dd-trace-go/pull/1042/commits/007399fa7b2f9deae49557e0366bba4634d3c67b

File: contrib/internal/httputil/trace.go
Comment (Julio-Guerra) [reply contrib/internal/httputil/trace.go]: Fixed in https://github.com/DataDog/dd-trace-go/pull/1042/commits/89a0faab579118328372ae6761d324707ccbad3c

File: contrib/labstack/echo.v4/echotrace.go
Comment (Julio-Guerra) [reply contrib/labstack/echo.v4/echotrace.go]: Fixed in https://github.com/DataDog/dd-trace-go/pull/1042/commits/89a0faab579118328372ae6761d324707ccbad3c

File: ddtrace/tracer/log.go
Comment (Julio-Guerra) [reply ddtrace/tracer/log.go]: Indeed, a bool value makes it simpler and good enough for debugging purposes.
Done in https://github.com/DataDog/dd-trace-go/pull/1042/commits/4f68efd71656021a57827d97dde5161a2e2a06e6

File: contrib/internal/httputil/trace.go
Comment (gbbr) [inline contrib/internal/httputil/trace.go:57]: I think you can remove this comment as it is absolutely obvious what is going on.

**Review by gbbr — APPROVED** [review]
Thanks Julio! This looks fine by me. If you want to wait for an additional review from @knusbaum, please do that.

**Review by knusbaum — APPROVED** [review]

File: contrib/internal/httputil/trace.go
Comment (Julio-Guerra) [reply contrib/internal/httputil/trace.go]: Fixed

**Review by knusbaum — APPROVED** [review]

**Pull request merged**
